File: templates/graph/advanced.go
package graph

func DetectCycleDirected(numNodes int, edges [][2]int) bool {
	graph := make([][]int, numNodes)
	for _, e := range edges {
		graph[e[0]] = append(graph[e[0]], e[1])
	}

	state := make([]int, numNodes)
	var dfs func(int) bool
	dfs = func(node int) bool {
		if state[node] == 1 {
			return true
		}
		if state[node] == 2 {
			return false
		}
		state[node] = 1
		for _, nei := range graph[node] {
			if dfs(nei) {
				return true
			}
		}
		state[node] = 2
		return false
	}

	for i := 0; i < numNodes; i++ {
		if state[i] == 0 && dfs(i) {
			return true
		}
	}
	return false
}

func MultiSourceBFS(grid [][]int, starts [][2]int) [][]int {
	rows, cols := len(grid), len(grid[0])
	dist := make([][]int, rows)
	for r := range dist {
		dist[r] = make([]int, cols)
		for c := range dist[r] {
			dist[r][c] = -1
		}
	}

	queue := make([][2]int, len(starts))
	copy(queue, starts)
	for _, p := range starts {
		dist[p[0]][p[1]] = 0
	}

	dirs := [][2]int{{1, 0}, {-1, 0}, {0, 1}, {0, -1}}
	for len(queue) > 0 {
		cur := queue[0]
		queue = queue[1:]
		for _, d := range dirs {
			nr, nc := cur[0]+d[0], cur[1]+d[1]
			if nr < 0 || nr >= rows || nc < 0 || nc >= cols || dist[nr][nc] != -1 {
				continue
			}
			dist[nr][nc] = dist[cur[0]][cur[1]] + 1
			queue = append(queue, [2]int{nr, nc})
		}
	}
	return dist
}

File: templates/graph/advanced.py
from collections import deque


def detect_cycle_directed(num_nodes: int, edges: list[tuple[int, int]]) -> bool:
    graph = [[] for _ in range(num_nodes)]
    for src, dst in edges:
        graph[src].append(dst)

    state = [0] * num_nodes

    def dfs(node: int) -> bool:
        if state[node] == 1:
            return True
        if state[node] == 2:
            return False
        state[node] = 1
        for nei in graph[node]:
            if dfs(nei):
                return True
        state[node] = 2
        return False

    return any(state[i] == 0 and dfs(i) for i in range(num_nodes))


def multi_source_bfs(grid: list[list[int]], starts: list[tuple[int, int]]) -> list[list[int]]:
    rows, cols = len(grid), len(grid[0])
    dist = [[-1] * cols for _ in range(rows)]
    queue = deque(starts)

    for row, col in starts:
        dist[row][col] = 0

    for row, col in queue:
        dist[row][col] = 0

    directions = ((1, 0), (-1, 0), (0, 1), (0, -1))
    while queue:
        row, col = queue.popleft()
        for dr, dc in directions:
            nr, nc = row + dr, col + dc
            if nr < 0 or nr >= rows or nc < 0 or nc >= cols or dist[nr][nc] != -1:
                continue
            dist[nr][nc] = dist[row][col] + 1
            queue.append((nr, nc))

    return dist
